### PR TITLE
feat(IamAssumeAuthenticator): introduce new authenticator type

### PR DESCRIPTION
**Pull request by padamstx**
This commit introduces the new IamAssumeAuthenticator which will fetch an IAM access token using the IAM getToken operation's "assume" grant type.  The resulting access token allows the application to assume the identity of a trusted profile, similar to the "sudo" feature of Linux.

### CODE REVIEW

File: src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java [hunk 1]
```diff
@@ -217,6 +231,15 @@ public void testFileCredentialsService4() {
     assertEquals(auth.authenticationType(), Authenticator.AUTHTYPE_NOAUTH);
   }
 
+  @Test
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java:234]: All I did with this code block is I moved it to be right after the test re: "service4".

File: src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java [hunk 2]
```diff
@@ -310,12 +333,48 @@ public void testFileCredentialsService17() {
   }
 
   @Test
-  public void testFileCredentialsService5() {
```
Comment (padamstx) [inline src/test/java/com/ibm/cloud/sdk/core/test/security/ConfigBasedAuthenticatorFactoryTest.java:313]: It's not totally obvious, but I moved "testFileCredentialsService5" up near "service4" and added three new test methods here.

**Review by pyrooka — APPROVED** [review]
Looks good! Just one thing I'd like to mention: in the `Authentication.md` I found 4 typos - outside of you changes. It's the `verificaton`. If you think you can fix those in this PR, totally up to you. :)

File: Authentication.md [hunk 3]
```diff
@@ -138,10 +139,10 @@ authenticator type is intended for situations in which the application will be m
 token itself in terms of initial acquisition and refreshing as needed.
 
 
-## Identity and Access Management (IAM) Authentication
-The `IamAuthenticator` will accept a user-supplied api key and will perform
+## Identity and Access Management (IAM) Authentication (grant type: apikey)
+The `IamAuthenticator` will accept a user-supplied apikey and will perform
 the necessary interactions with the IAM token service to obtain a suitable
-bearer token for the specified api key.  The authenticator will also obtain 
+bearer token for the specified apikey.  The authenticator will also obtain
```
[the remark this answers is not in the document]
Comment (padamstx) [reply Authentication.md]: fixed

**Comment by padamstx** [comment]
> in the `Authentication.md` I found 4 typos

fixed

**Review by dpopp07 — APPROVED** [review]
Looks good! 👍

**Comment by padamstx** [comment]
Just pushed the commit with the immutable-related changes.  I'll request a re-review...

**Review by dpopp07 — APPROVED** [review]
New changes look good!

**Review by pyrooka — APPROVED** [review]
LGTM!

**Pull request merged**